src/ProxyClient/AbstractProxyClient.php

```diff
@@ -61,58 +52,16 @@
      *                                 requests (optional). This is required if
      *                                 you purge and refresh paths instead of
      *                                 absolute URLs.
-     * @param ClientInterface $client  HTTP client (optional). If no HTTP client
-     *                                 is supplied, a default one will be
-     *                                 created.
-     */
-    public function __construct(array $servers, $baseUrl = null, ClientInterface $client = null)
-    {
-        $this->client = $client ?: new Client();
-        $this->setServers($servers);
-        $this->setBaseUrl($baseUrl);
-    }
-
-    /**
-     * Set caching proxy servers
-     *
-     * @param array $servers Caching proxy proxy server hostnames or IP
-     *                       addresses, including port if not port 80.
-     *                       E.g. array('127.0.0.1:6081')
-     *
-     * @throws InvalidUrlException If server is invalid or contains URL
-     *                             parts other than scheme, host, port
-     */
-    public function setServers(array $servers)
-    {
-        $this->servers = array();
-        foreach ($servers as $server) {
-            $this->servers[] = $this->filterUrl($server, array('scheme', 'host', 'port'));
-        }
-    }
-
-    /**
-     * Set application hostname, optionally including a base URL, for purge and
-     * refresh requests
-     *
-     * @param string $url Your application’s base URL or hostname
-     */
-    public function setBaseUrl($url)
-    {
-        if ($url) {
-            $url = $this->filterUrl($url);
-        }
-
-        $this->client->setBaseUrl($url);
-    }
-
-    /**
-     * Get application base URL
-     *
-     * @return string Your application base url
-     */
-    protected function getBaseUrl()
-    {
-        $this->client->getBaseUrl();
+     * @param PsrHttpAdapter $httpAdapter If no HTTP adapter is supplied, a
+     *                                 default one will be created.
+     */
+    public function __construct(
+        array $servers,
+        $baseUrl = null,
+        PsrHttpAdapter $httpAdapter = null
+    ) {
+        $this->httpAdapter = $httpAdapter ?: new Guzzle6HttpAdapter();
```

Comment:
Actually, you can handle this with composer. Each adapter implementation provides a virtual package: `php-http/adapter-implementation`. You can require this virtual package instead of a concrete implementation. The factory should find all available clients and choose the best (which depends on personal preference, but lets say guzzle is the best and always use that if available)